### DISCUSSION TITLE
Adding ip_hdr->ttl to conn.log

### DISCUSSION THREAD

**Discussion by mintos5** [zeek/zeek#2495] [Archive]
Hi all,

i have difficult task of monitoring **Time to live** field from ip header in all of our traffic.
I want to get the the TTL value from the first packet of Connection, other packets can be ignored.
It is not by default in there, but you can easily add information from layer2 like mac addresses with this policy:
https://github.com/zeek/zeek/blob/master/scripts/policy/protocols/conn/mac-logging.zeek

So how hard can it be? I found out, that even connection record data type doesn't have it or even pointer to packet_header.
So I made this little hack, that adds needed field first to connection record and then to Conn::Info record that is written to **conn.log**

Here is the diff needed for version 5.0.2:
```diff
diff --git scripts/base/init-bare.zeek scripts/base/init-bare.zeek
index 46d2bc867..9a40c4df9 100644
--- scripts/base/init-bare.zeek
+++ scripts/base/init-bare.zeek
@@ -484,6 +484,7 @@ type connection: record {
 
 	## The inner VLAN, if applicable for this connection.
 	inner_vlan: int &optional;
+	ttl: count &optional;
 };
 
 ## Arguments given to Zeek from the command line. In order to use this, Zeek
diff --git scripts/base/protocols/conn/main.zeek scripts/base/protocols/conn/main.zeek
index cda5261bf..71f309e51 100644
--- scripts/base/protocols/conn/main.zeek
+++ scripts/base/protocols/conn/main.zeek
@@ -148,6 +148,7 @@ export {
 		## *uid* values for any encapsulating parent connections
 		## used over the lifetime of this inner connection.
 		tunnel_parents: set[string] &log &optional;
+		first_ttl: count				&log &optional;
 	};
 
 	## Event that can be handled to access the :zeek:type:`Conn::Info`
@@ -245,6 +246,10 @@ function set_conn(c: connection, eoc: bool)
 		}
 
 	c$conn$id=c$id;
+	if ( c?$ttl )
+		{
+			c$conn$first_ttl=c$ttl;
+		}
 	if ( c?$tunnel && |c$tunnel| > 0 )
 		{
 		if ( ! c$conn?$tunnel_parents )
diff --git src/Conn.cc src/Conn.cc
index 59052c2cb..7e9249b36 100644
--- src/Conn.cc
+++ src/Conn.cc
@@ -57,6 +57,11 @@ Connection::Connection(const detail::ConnKey& k, double t, const ConnTuple* id,
 	vlan = pkt->vlan;
 	inner_vlan = pkt->inner_vlan;
 
+	if (pkt->ip_hdr)
+		ttl = pkt->ip_hdr->TTL();
+	else
+		ttl = NULL;
+
 	weird = 0;
 
 	suppress_event = 0;
@@ -270,6 +275,8 @@ const RecordValPtr& Connection::GetVal()
 
 		if ( inner_vlan != 0 )
 			conn_val->Assign(10, inner_vlan);
+		if ( ttl != 0 )
+			conn_val->Assign(11, ttl);
 		}
 
 	if ( adapter )
diff --git src/Conn.h src/Conn.h
index ca7c0eeeb..add905f5e 100644
--- src/Conn.h
+++ src/Conn.h
@@ -260,6 +260,7 @@ private:
 	TransportProto proto;
 	uint32_t orig_flow_label, resp_flow_label; // most recent IPv6 flow labels
 	uint32_t vlan, inner_vlan; // VLAN this connection traverses, if available
+	uint8_t ttl;  //experimental addition of ttl to conn
 	u_char orig_l2_addr[Packet::L2_ADDR_LEN]; // Link-layer originator address, if available
 	u_char resp_l2_addr[Packet::L2_ADDR_LEN]; // Link-layer responder address, if available
 	int suppress_event; // suppress certain events to once per conn.
```

I don't like this hacky way of adding this field to our logs so I am thinking of other ways to get the same results?

Maybe using event **new_packet**? From there I can acces pkt_hdr->ipv4_hdr->ttl
https://docs.zeek.org/en/master/scripts/base/bif/event.bif.zeek.html#id-new_packet
But the docs are saying: This is a very low-level and expensive event that should be avoided when at all possible.
So can I use it? I am not sure and it looks like it will be fired for every packet and I don't need that.

Other option I think is to write own plugin that will be implementing Packet Analyzer API. But do I want to create transport-layer analyzer, that is registered by ports (_PacketAnalyzer::register_for_ports_) or other type Packet Analyzer?
I am little bit lost with writing a plugin for Zeek.

Hoping for ideas and suggestion how to do it the best way.

**Comment by awelzel** [comment]
Hey @mintos5 - interesting!

> I don't like this hacky way of adding this field to our logs so I am thinking of other ways to get the same results?

Is the following something that could possibly work for you without needing to modify the core or adding an extra analyzer?
```
# Need access to Conn::set_conn() to populate c$conn,
# so do it in the Conn module, else would need to stash ttl
# somewhere else.
module Conn;

redef record Conn::Info += {
        ttl: count &optional &log;
};

event new_connection(c: connection) {
        Conn::set_conn(c, F);
        local hdr = get_current_packet_header();

        if ( hdr?$ip )
                c$conn$ttl = hdr$ip$ttl;
}
```
Calling `get_current_packet_header()` within `new_connection()` should give the first packet header in a connection that Zeek tracks and it seems to do what you're looking for.

One caveat here is that the packet header from `get_current_packet_header()` represents the most outer header. This becomes relevant if your interested in the `ttl` of encapsulated packets instead (think IP packets encapsulated within ipip/vxlan/geneve tunnels). If you don't have tunnels or decapsulate traffic before passing it to Zeek, above script might be an alternative to changes in the core or adding an analyzer.

I do wonder if there's a case for adding `get_current_inner_packet_header()` and `get_current_inner_packet()` (or something more thought through) to cover/enable encapsulation cases? We might be tracking the "right packet" in `run_state::current_pkt` on the C++ side already.

**Reply by mintos5** [reply]
Hi @awelzel, thank You for zeek script abowe. It is working well for me.
We are not analyzing encapsulated packets, so we are getting needed correct ttl value.

In the future I will try to search more for beneficial BIF functions than hacking/modifying core ;)
